File: 2024/golang/src/day1/main.go
package main

import (
	"math"
	"sort"
	"strconv"
	"strings"
	"utils"
)

func GetData(filename string) interface{} {
	lines := utils.GetLines(filename)
	var l1 []int
	var l2 []int
	for _, line := range lines {
		parts := strings.Split(line, "   ")
		atoi, _ := strconv.Atoi(parts[0])
		l1 = append(l1, atoi)

		atoi, _ = strconv.Atoi(parts[1])
		l2 = append(l2, atoi)
	}
	sort.Slice(l1, func(i, j int) bool {
		return l1[i] < l1[j]
	})
	sort.Slice(l2, func(i, j int) bool {
		return l2[i] < l2[j]
	})
	return [][]int{l1, l2}
}

func PartOne(d interface{}) string {
	data := d.([][]int)

	s := 0
	for i, num := range data[0] {
		s += int(math.Abs(float64(num - data[1][i])))
	}
	return strconv.Itoa(s)
}

func PartTwo(d interface{}) string {
	data := d.([][]int)

	counts := make(map[int]int)
	for _, num := range data[1] {
		counts[num]++
	}

	s := 0
	for _, num := range data[0] {
		s += num * counts[num]
	}
	return strconv.Itoa(s)
}

func main() {
	data := GetData("input.txt")
	utils.RunWithTimer(PartOne, data)
	utils.RunWithTimer(PartTwo, data)
}

File: 2024/python/day1/main.py
from utils.timers import run_with_timer, get_data_with_timer
from collections import defaultdict


def get_data(filename):
	lines = [list(map(int, x.strip().split())) for x in open(filename).readlines()]
	l1 = [x[0] for x in lines]
	l2 = [x[1] for x in lines]
	l1.sort()
	l2.sort()
	return [l1, l2]


def part_one(d):
	return sum(abs(d[0][x] - d[1][x]) for x in range(len(d[0])))


def part_two(d):
	counts = defaultdict(lambda: 0)
	for x in d[1]:
		counts[x] += 1
	return sum(x * (counts[x] if x in counts else 0) for x in d[0])


def main(f="input.txt"):
	data = get_data_with_timer(get_data, f)
	run_with_timer(part_one, data)
	run_with_timer(part_two, data)


if __name__ == '__main__':
	main()
